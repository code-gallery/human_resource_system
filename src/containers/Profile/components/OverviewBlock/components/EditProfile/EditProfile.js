import React from 'react'
import PropTypes from 'prop-types'
import Datetime from 'react-datetime'
import Autocomplete from 'react-google-autocomplete'
import _isNil from 'lodash/isNil'
import moment from 'moment'
import { renderValidation } from 'containers/Profile/utils/validation'
import TwitterInput from '../TwitterInput'

const EditProfile = (props) => {
  const {
    onChangeProfileField,
    invalidFields,
    user: {
      first_name,
      last_name,
      twitter,
      tagline,
      public_email,
      mobile,
      town,
      country,
      dob
    }
  } = props
  const locationPlaceholder = ((_isNil(town) || town === '') && (_isNil(country) || country === '')) ?
    'Enter your address' : `${town}, ${country}`
  const dateValue = (dob) ? moment(dob) : ''

  return (
    <div className="profile-overview-edit">
      <p className="profile-block-title">Details</p>
      <div className="form-group">
        <label className={renderValidation(invalidFields, 'first_name')}>First Name</label>
        <input
          name="first_name"
          type="text"
          defaultValue={first_name}
          onChange={e => onChangeProfileField('first_name', e.target.value)}
        />
      </div>
      <div className="form-group">
        <label className={renderValidation(invalidFields, 'last_name')}>Last Name</label>
        <input
          name="last_name"
          type="text"
          defaultValue={last_name}
          onChange={e => onChangeProfileField('last_name', e.target.value)}
        />
      </div>
      <div className="form-group">
        <label>About</label>
        <textarea
          name="tagline"
          value={tagline || ''}
          onChange={e => onChangeProfileField('tagline', e.target.value)}
        />
      </div>
      <div className="form-group">
        <label>Location</label>
        <Autocomplete
          onPlaceSelected={res => onChangeProfileField('location', res)}
          placeholder={locationPlaceholder}
          types={[ '(cities)' ]}
        />
      </div>
      <div className="form-group">
        <label>Twitter</label>
        <TwitterInput
          value={twitter}
          onChangeInput={val => onChangeProfileField('twitter', val)} />
      </div>
      <div className="form-group">
        <label>Public Email</label>
        <input
          name="public_email"
          type="text"
          defaultValue={public_email}
          onChange={e => onChangeProfileField('public_email', e.target.value)}
        />
      </div>
      <div className="form-group">
        <label>Public Telephone</label>
        <input
          name="mobile"
          type="text"
          defaultValue={mobile}
          onChange={e => onChangeProfileField('mobile', e.target.value)}
        />
      </div>
      <div className="form-group">
        <label>Date of Birth</label>
        <Datetime
          name="dob"
          value={dateValue}
          dateFormat="DD/MM/YYYY"
          timeFormat={false}
          onChange={(date) => {
            if (date && date.format) {
              onChangeProfileField('dob', date.format('YYYY-MM-DD'))
            }
          }}
        />
      </div>
    </div>
  )
}

EditProfile.propTypes = {
  user: PropTypes.object,
  onChangeProfileField: PropTypes.func.isRequired,
  invalidFields: PropTypes.array.isRequired
}

export default EditProfile
