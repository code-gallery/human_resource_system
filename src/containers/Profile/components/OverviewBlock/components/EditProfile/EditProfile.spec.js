import React from 'react'
import renderer from 'react-test-renderer'
import { shallow } from 'enzyme'
import EditProfile from './index'

describe('containers/Profile/components/.../EditProfile', () => {
  describe('with no validation errors', () => {
    it('renders correctly', () => {
      const props = {
        user: {
          first_name: 'Jane',
          last_name: 'Doe'
        },
        onChangeProfileField: jest.fn(),
        invalidFields: []
      }
      const tree = renderer.create(
        <EditProfile {...props} />
      ).toJSON()

      expect(tree).toMatchSnapshot()
    })
  })

  describe('with validation errors', () => {
    it('renders correctly', () => {
      const props = {
        user: {
          first_name: '',
          last_name: ''
        },
        onChangeProfileField: jest.fn(),
        invalidFields: [ 'first_name', 'last_name' ]
      }
      const tree = renderer.create(
        <EditProfile {...props} />
      ).toJSON()

      expect(tree).toMatchSnapshot()
    })
  })

  describe('<EditProfile />', () => {
    let wrapper
    let props
    beforeEach(() => {
      props = {
        user: {},
        onChangeProfileField: jest.fn(),
        invalidFields: []
      }
      wrapper = shallow(
        <EditProfile {...props} />
      )
    })

    describe('when "first_name" changes', () => {
      it('calls props.onChangeProfileField correctly', () => {
        const event = { target: { type: 'text', value: 'Jane' } }
        wrapper.find('[name="first_name"]').simulate('change', event)
        expect(props.onChangeProfileField).toHaveBeenCalledWith('first_name', 'Jane')
      })
    })

    describe('when "last_name" changes', () => {
      it('calls props.onChangeProfileField correctly', () => {
        const event = { target: { type: 'text', value: 'Doe' } }
        wrapper.find('[name="last_name"]').simulate('change', event)
        expect(props.onChangeProfileField).toHaveBeenCalledWith('last_name', 'Doe')
      })
    })

    describe('when "tagline" changes', () => {
      it('calls props.onChangeProfileField correctly', () => {
        const event = { target: { type: 'text', value: 'Lorem ipsum dolor' } }
        wrapper.find('[name="tagline"]').simulate('change', event)
        expect(props.onChangeProfileField).toHaveBeenCalledWith('tagline', 'Lorem ipsum dolor')
      })
    })

    describe('when "public_email" changes', () => {
      it('calls props.onChangeProfileField correctly', () => {
        const event = { target: { type: 'text', value: 'ab@appliedblockchain.com' } }
        wrapper.find('[name="public_email"]').simulate('change', event)
        expect(props.onChangeProfileField).toHaveBeenCalledWith('public_email', 'ab@appliedblockchain.com')
      })
    })

    describe('when "mobile" changes', () => {
      it('calls props.onChangeProfileField correctly', () => {
        const event = { target: { type: 'text', value: '077000' } }
        wrapper.find('[name="mobile"]').simulate('change', event)
        expect(props.onChangeProfileField).toHaveBeenCalledWith('mobile', '077000')
      })
    })

    describe('when "dob" changes', () => {
      it('calls props.onChangeProfileField correctly', () => {
        const event = { format: function() {
          return '21/09/2016'
        } }
        wrapper.find('[name="dob"]').simulate('change', event)
        expect(props.onChangeProfileField).toHaveBeenCalledWith('dob', '21/09/2016')
      })
    })
  })
})
