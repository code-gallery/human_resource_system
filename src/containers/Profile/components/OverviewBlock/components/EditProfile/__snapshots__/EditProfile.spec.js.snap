// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`containers/Profile/components/.../EditProfile with no validation errors renders correctly 1`] = `
<div
  className="profile-overview-edit"
>
  <p
    className="profile-block-title"
  >
    Details
  </p>
  <div
    className="form-group"
  >
    <label
      className=""
    >
      First Name
    </label>
    <input
      defaultValue="Jane"
      name="first_name"
      onChange={[Function]}
      type="text"
    />
  </div>
  <div
    className="form-group"
  >
    <label
      className=""
    >
      Last Name
    </label>
    <input
      defaultValue="Doe"
      name="last_name"
      onChange={[Function]}
      type="text"
    />
  </div>
  <div
    className="form-group"
  >
    <label>
      About
    </label>
    <textarea
      name="tagline"
      onChange={[Function]}
      value=""
    />
  </div>
  <div
    className="form-group"
  >
    <label>
      Location
    </label>
    <div>
      <input
        placeholder="Enter your address"
      />
    </div>
  </div>
  <div
    className="form-group"
  >
    <label>
      Twitter
    </label>
    <input
      onChange={[Function]}
      type="text"
      value=""
    />
  </div>
  <div
    className="form-group"
  >
    <label>
      Public Email
    </label>
    <input
      defaultValue={undefined}
      name="public_email"
      onChange={[Function]}
      type="text"
    />
  </div>
  <div
    className="form-group"
  >
    <label>
      Public Telephone
    </label>
    <input
      defaultValue={undefined}
      name="mobile"
      onChange={[Function]}
      type="text"
    />
  </div>
  <div
    className="form-group"
  >
    <label>
      Date of Birth
    </label>
    <DateTimeMock
      dateFormat="DD/MM/YYYY"
      name="dob"
      onChange={[Function]}
      timeFormat={false}
      value=""
    />
  </div>
</div>
`;

exports[`containers/Profile/components/.../EditProfile with validation errors renders correctly 1`] = `
<div
  className="profile-overview-edit"
>
  <p
    className="profile-block-title"
  >
    Details
  </p>
  <div
    className="form-group"
  >
    <label
      className="required-error"
    >
      First Name
    </label>
    <input
      defaultValue=""
      name="first_name"
      onChange={[Function]}
      type="text"
    />
  </div>
  <div
    className="form-group"
  >
    <label
      className="required-error"
    >
      Last Name
    </label>
    <input
      defaultValue=""
      name="last_name"
      onChange={[Function]}
      type="text"
    />
  </div>
  <div
    className="form-group"
  >
    <label>
      About
    </label>
    <textarea
      name="tagline"
      onChange={[Function]}
      value=""
    />
  </div>
  <div
    className="form-group"
  >
    <label>
      Location
    </label>
    <div>
      <input
        placeholder="Enter your address"
      />
    </div>
  </div>
  <div
    className="form-group"
  >
    <label>
      Twitter
    </label>
    <input
      onChange={[Function]}
      type="text"
      value=""
    />
  </div>
  <div
    className="form-group"
  >
    <label>
      Public Email
    </label>
    <input
      defaultValue={undefined}
      name="public_email"
      onChange={[Function]}
      type="text"
    />
  </div>
  <div
    className="form-group"
  >
    <label>
      Public Telephone
    </label>
    <input
      defaultValue={undefined}
      name="mobile"
      onChange={[Function]}
      type="text"
    />
  </div>
  <div
    className="form-group"
  >
    <label>
      Date of Birth
    </label>
    <DateTimeMock
      dateFormat="DD/MM/YYYY"
      name="dob"
      onChange={[Function]}
      timeFormat={false}
      value=""
    />
  </div>
</div>
`;
